Foundation/NSDecimal.swift

```diff
@@ -56,74 +187,286 @@ extension Decimal {
 }
 
 extension Decimal : Hashable, Comparable {
-    public var hashValue: Int { NSUnimplemented() }
-    public static func ==(lhs: Decimal, rhs: Decimal) -> Bool { NSUnimplemented() }
-    public static func <(lhs: Decimal, rhs: Decimal) -> Bool { NSUnimplemented() }
+    internal var doubleValue: Double {
+        var d = 0.0
+        if _length == 0 && _isNegative == 0 {
+            return Double.nan
+        }
+        for i in 0..<8 {
+            let index = 8 - i - 1
+            switch index {
+            case 0:
+                d = d * 65536 + Double(_mantissa.0)
+                break
+            case 1:
+                d = d * 65536 + Double(_mantissa.1)
+                break
+            case 2:
+                d = d * 65536 + Double(_mantissa.2)
+                break
+            case 3:
+                d = d * 65536 + Double(_mantissa.3)
+                break
+            case 4:
+                d = d * 65536 + Double(_mantissa.4)
+                break
+            case 5:
+                d = d * 65536 + Double(_mantissa.5)
+                break
+            case 6:
+                d = d * 65536 + Double(_mantissa.6)
+                break
+            case 7:
+                d = d * 65536 + Double(_mantissa.7)
+                break
+            default:
+                fatalError("conversion overflow")
+            }
+        }
+        if _exponent < 0 {
+            for _ in _exponent..<0 {
+                d /= 10.0
+            }
+        } else {
+            for _ in 0..<_exponent {
+                d *= 10.0
+            }
+        }
+        return _isNegative != 0 ? -d : d
+    }
+    public var hashValue: Int {
+        return Int(bitPattern: __CFHashDouble(doubleValue))
+    }
+    public static func ==(lhs: Decimal, rhs: Decimal) -> Bool {
+        if lhs.isNaN || rhs.isNaN {
+            return false
```

Comment:
To match the behavior of the objective-c version:
```objc
    // NaN is a special case and is arbitrary ordered before everything else
    // Conceptually comparing with NaN is bogus anyway but raising or 
    // always returning the same answer will confuse the sorting algorithms
   if(NSDecimalIsNotANumber(leftOperand)) {
        if(NSDecimalIsNotANumber(rightOperand)) {
            return NSOrderedSame;
        }
        return NSOrderedAscending;
    }
```

so if both are NaN then they are ordered same which is considered equal (irrespective of any other behavior standard of floating point numbers)